fix: #1490: Make pfe-icon set CSS properties instead of overriding local var

wesruv opened the pull request:
Overriding local vars is overriding author's intentions in some cases, making pfe-icon set the CSS properties the vars effect instead of overriding the vars

<!-- Reminder: please DO NOT forget your changelog and visual baseline updates! -->

## <Pull request description>


### Related issues

- (#1490) PFE-Icon is setting local variables overriding author's intentions


### Preview

Link(s) to demo page(s) where this element can be viewed:
- [Icon demo](https://deploy-preview-1491--happy-galileo-ea79c4.netlify.app/elements/pfe-icon/demo/) 


### Testing instructions

1. Go to pfe-icon demo page and make sure nothing has changed, but setting --pfe-icon-Color should change icon color in all circumstances.


#### Browser requirements

Your component should work in all of the following environments:

- [x] Latest 2 versions of Edge
- [x] Internet Explorer 11 (should be useable, not pixel perfect)
- [x] Latest 2 versions of Firefox (one on Mac OS, one of Windows OS)
- [x] Firefox 68 (or latest version for Red Hat Enterprise Linux distribution)
- [x] Latest 2 versions of Chrome (one on Mac OS, one of Windows OS)
- [x] Latest 2 versions of Safari
- [x] Android mobile device (such as the Galaxy S9)
- [x] Apple mobile device (such as the iPhone X)
- [x] Apple tablet device (such as the iPhone Pro)


### Ready-for-merge Checklist

- [x] Expected files: all files in this pull request are related to one request or issue (no stragglers or scope-creep).
- [ ] Browser testing passed.
- [x] Repository compiles and tests pass.
- [x] Changelog updated.


### Merging

Please **squash** when merging and ensure your commit message uses [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) formatting.

**Be sure to share your updates with the [patternfly-elements-contribute@redhat.com](mailto:patternfly-elements-contribute@redhat.com) mailing list!**

castastrophe commented on elements/pfe-icon/src/pfe-icon.scss:
```diff
@@ -113,7 +113,9 @@ $sizing: (
   :host([color="#{$level}"]) {
-    --pfe-icon--color: #{pfe-var(feedback--#{$level})};
+    filter feFlood {
+      flood-color: #{pfe-var(feedback--#{$level})};
```
To prevent regressions, I'd recommend:

```suggestion
      flood-color: pfe-local(color, $fallback: pfe-var(feedback--#{$level}));
```

Also just a note that you don't need interpolation on functions when assigning them to css properties, only when assigning them to custom properties.

castastrophe commented on elements/pfe-icon/src/pfe-icon.scss:
```diff
@@ -130,11 +132,14 @@ $sizing: (
+  background-color: pfe-local(BackgroundColor, pfe-local(color, $fallback: pfe-var(surface--lightest)));
+
+  filter feFlood {
+    flood-color: pfe-local(Color, $fallback: pfe-broadcasted(text));
```
```suggestion
    flood-color: pfe-local(color, $fallback: pfe-broadcasted(text));
```

This component is using a lowercase `color` instead because it's the concept of color rather than the literal `color` property in CSS.

wesruv replied:
Finding that very confusing, it doesn't seem to be documented in the code or the README, is that something someone who understands it can write down the differences between the vars?

castastrophe replied:
Yeah, now that Kyle is close to merging the documentation update, Kendall & I are working on bringing over the write-up but the gist is BEM.  In BEM, a property that is specific to a CSS property uses this format: BackgroundColor vs. background-color or backgroundColor.  It's how you can separate the idea of a color variable from a variable that literally is assigned to the color property in CSS.  Since this variable is assigned to flood-color, the options were: FloodColor or a more abstract naming.  Since you don't really need to know how the inner workings of the SVG function to style this component, `--pfe-icon--color` was chosen.

It will all be written out but for now, the summary is: `--pfe-icon--color` is correct over `--pfe-icon--Color`.

castastrophe replied:
It's also in the README: 

```
For more fine-grained control, `--pfe-icon--color` is available to override the color of a specific icon or sets of icons and will be applied to either the SVG lines or the background of the circle (if circled).
```

wesruv replied:
Oh, damn. I thought it was, but then I checked and skimmed over it. 😵 Resolving convo

castastrophe reviewed (changes requested):
The code looks good, just a couple small updates.  Needs a changelog added as well.  Thanks!

wesruv commented:
Rut-roh, pretty sure this isn't related to this PR, can anyone else duplicate this:

My PR:
![image](https://user-images.githubusercontent.com/5607236/112519409-b952d880-8d70-11eb-9b2a-403e83a0201f.png)

PatternflyElements.com
![image](https://user-images.githubusercontent.com/5607236/112519427-bfe15000-8d70-11eb-9f57-a58ac2196378.png)

wesruv commented:
Well, this PR hasn't caused any regressions at least 😬 

![image](https://user-images.githubusercontent.com/5607236/112519921-3bdb9800-8d71-11eb-8e6d-e4683139957c.png)

![image](https://user-images.githubusercontent.com/5607236/112519937-4138e280-8d71-11eb-92cc-e9115c1aea19.png)

castastrophe commented:
@wesruv What environment?  IE11?  Edge?  I know that patternflyelements.com won't work in IE11 because it uses module imports but the icon demo page should: https://patternflyelements.com/elements/pfe-icon/demo/

This is Chrome on Mac:
<img width="1487" alt="Screen Shot 2021-03-25 at 1 53 38 PM" src="https://user-images.githubusercontent.com/1840295/112520188-8bba5f00-8d71-11eb-9604-051b3a1619f9.png">

castastrophe commented:
Something's up with the colors in Chrome though.  This is https://deploy-preview-1491--happy-galileo-ea79c4.netlify.app/elements/pfe-icon/demo/ on Chrome Mac:
<img width="1494" alt="Screen Shot 2021-03-25 at 1 55 11 PM" src="https://user-images.githubusercontent.com/1840295/112520409-c3c1a200-8d71-11eb-8da1-28b4f806dd6a.png">

castastrophe commented:
This section might need to be updated as well to not use the `--pfe-icon--color` override approach: https://github.com/patternfly/patternfly-elements/blob/1987edcd2bfa20cdfd40754b0de2e9e879ae0514/elements/pfe-icon/src/pfe-icon.scss#L127

wesruv commented:
Getting the same issues in browser-stack :(

![image](https://user-images.githubusercontent.com/5607236/112521103-814c9500-8d72-11eb-9328-097da5b5dd90.png)

wesruv commented:
> This section might need to be updated as well to not use the `--pfe-icon--color` override approach:
> 
> https://github.com/patternfly/patternfly-elements/blob/1987edcd2bfa20cdfd40754b0de2e9e879ae0514/elements/pfe-icon/src/pfe-icon.scss#L127

I see it on Linux too, I'm on it

castastrophe commented on elements/pfe-icon/src/pfe-icon.scss:
```diff
@@ -107,13 +107,23 @@ $sizing: (
-    --pfe-icon--size: #{$size};
+    // Respect var if user set CSS var, but fallback to sizing property map
+    $desired-size: pfe-local(size, $fallback: $size);
+    max-width:   calc(#{$desired-size} + #{pfe-local(Padding)} * 2);
```
This looks really great, I love the addition of the $desired-color and $desired-size to create that shared source of truth.  Can you tell me more about the addition of padding into the width calculation?

castastrophe replied:
I'm wondering also if, without breaking it out into `PaddingTop`, `PaddingRight` etc. if this will be an issue.  Imagine I've done: `--pfe-icon--Padding: 10px 5px;` in my stylesheet.  How would that work inside this calc?

wesruv replied:
I'm just maintaining what was happening when the variables were updated (which turned out to be much more complicated than I expected). If you look here: https://github.com/patternfly/patternfly-elements/blob/b8a3c6e8ba0e5e97286c9765fee9159f51bd15f0/elements/pfe-icon/src/pfe-icon.scss#L40

That's what I'm maintaining, I think it's because the width is set to the SVG, and the max-width/height are set on the host, so the host has to take the padding into account. I thought about messing with it to see if I could get `box-sizing` or something a little more straightforward to make the same result, but I'm thinking it wouldn't work and I was having enough issues updating the SCSS and not getting a regression, so I just maintained what was there.

castastrophe replied:
Awesome!  Thanks for clarifying.  I'll add an issue to re-address that later.

castastrophe replied:
https://github.com/patternfly/patternfly-elements/issues/1502

castastrophe commented on elements/pfe-icon/demo/index.html:
```diff
@@ -125,8 +125,8 @@
     body {
-      --pfe-icon--color: orange;
-      --pfe-icon--context: light;
+      /* --pfe-icon--color: orange;
```
```suggestion
      /* Icon overrides for testing */
      /* --pfe-icon--color: orange;
```

^ A comment here might be helpful for future investigators

castastrophe reviewed (commented):
Looks great!  Just needs a changelog

wesruv commented:
I updated the Changelog, hopefully did that well?

I re-tested everything in IE, and for some reason now we're getting black icons, where before they were blank. Going to consider that a win?
![image](https://user-images.githubusercontent.com/5607236/113326758-178b3880-92e8-11eb-88af-b8f906218a73.png)

wesruv commented:
Changelog in, retested everything in the different browsers, and got some IE weirdness.

So I deleted a comment about IE, had a comment about IE that icons weren't showing up. Somehow I think taking a screenshot got me icons? Maybe forced a paint?? I swear they weren't there right before I screencapped.

I'm seeing pfe-icon render much slower in IE on patternflyelements.com consistently, but it seems to be ok on my feature branch? No idea why that would be.

Left is feature branch, right is current state:
![image](https://user-images.githubusercontent.com/5607236/113327131-91bbbd00-92e8-11eb-8da8-2103b193a8d6.png)

castastrophe commented:
This component is definitely a bit odd in IE as-is but I think that's a-okay.  Since it's icons, it's not critical that it upgrade fast or perfectly in IE.

castastrophe commented:
Re: changelog.  The one we need to keep updated is actually the CHANGELOG-1.x.md on the root of the project.  We talked about using local changelogs at one point and aggregating that up into a shared one but... nothing came from that.  So for now, no component-specific changelogs, just adding our entries to the CHANGELOG-1.x.md.

castastrophe commented:
Can you take a look at the failing tests when you get a chance? It looks like pfe-icon is getting black instead of the colors but I suspect it's how the styles are being queried in the test.

https://github.com/patternfly/patternfly-elements/pull/1491/checks?check_run_id=2247922228

wesruv commented:
Everything should be good to go, I updated an existing test that checks color change to check `--pfe-icon--color` and added a copy of that test to check if the legacy var `--pfe-icon--Color` works and updated the verbiage so it's more clear what's being tested.

castastrophe commented on elements/pfe-icon/CHANGELOG.md:
```diff
@@ -0,0 +1,5 @@
+## 1.3.4 ( TBD )
```
This can be deleted

castastrophe commented on elements/pfe-icon/test/index.html:
```diff
@@ -12,12 +12,12 @@
       'pfe-icon_test.html',
-      'pfe-icon_react_test.html',
-      'pfe-icon_vue_test.html',
+      // 'pfe-icon_react_test.html',
```
The other tests suites are still commented out

wesruv replied:
OH CRAP, good catch.

Sorry, I was debugging something else and forgot to take this back 😬

wesruv commented:
Ok, noooooow everything should be good to go?

castastrophe commented:
In running the visual snapshots, I'm seeing a degredation in the fallback for IE11 in the forced stretch section.

Current branch:
<img width="1354" alt="Screen Shot 2021-04-08 at 11 47 18 AM" src="https://user-images.githubusercontent.com/1840295/114057146-52e5b400-9860-11eb-8b33-1f654ddd37df.png">

Main branch:
<img width="772" alt="Screen Shot 2021-04-08 at 11 47 33 AM" src="https://user-images.githubusercontent.com/1840295/114057207-5da04900-9860-11eb-96f3-c808c1508aee.png">

wesruv commented:
Thaaaat's weird, it's not doing that on my Win 10 IE11:
![image](https://user-images.githubusercontent.com/5607236/114058584-98ef4780-9861-11eb-9ee8-ab799b724c27.png)

castastrophe commented:
Try opening the `/elements/pfe-icon/demo/index_e2e.html` page.  We snapshot that one for visual regression tests rather than the index.html file.

wesruv commented:
Got it! Thanks :D

wesruv commented:
For posterity, it looks like IE11 hates a unitless length in a `calc()` in this case it was `calc(64px + 0 * 2)`, `0` being the calculated CSS property of the padding.

_Fortunately_ I was able to switch the `:host` to `box-sizing: content-box` and that made that calc unecessary, w00t! Seems like using that approach is just better? I added `!important` because grid styles in the demo were making it `border-box` and that's a pretty common light DOM style.

castastrophe commented on elements/pfe-icon/src/pfe-icon.scss:
```diff
@@ -31,15 +38,16 @@ $LOCAL-VARIABLES: map-deep-merge($LOCAL-VARIABLES, (
   position: relative;
   display:  inline-block;
 
   // Supports sizing while preserving aspect ratio (despite layout)
```
Added the additional bit about the content-box model since that is an important part of making this work and we want to make sure no one accidentally removes it in the future.
